### PR TITLE
AO3-6571 Prevent relative image paths.

### PR DESCRIPTION
**Pull request by tickinginstant**
## Issue

https://otwarchive.atlassian.net/browse/AO3-6571

## Purpose

1. Adds a transformer that looks for `` elements, and uses `URL.join` on their `src` value to convert relative paths to absolute paths.
2. Modifies the sanitizer config to prevent relative paths in the `` `src` attribute, in case anything slips through the transformer.
3. Bumps the sanitizer version, so that all existing fields will be re-sanitized.

### CODE REVIEW

File: config/initializers/gem-plugin_config/sanitizer_config.rb [hunk 1]
```diff
@@ -33,7 +33,7 @@ module Config
       protocols: {
         "a" => { "href" => ["ftp", "http", "https", "mailto", :relative] },
         "blockquote" => { "cite" => ["http", "https", :relative] },
-        "img" => { "src" => ["http", "https", :relative] },
+        "img" => { "src" => ["http", "https"] },
```
[the remark this answers is not in the document]
Comment (tickinginstant) [reply config/initializers/gem-plugin_config/sanitizer_config.rb:36]: Hmm. I see Hound's point, but I also like the parallel construction between this and the nearby lines above and below, which contain a mixture of symbols and strings and therefore can't be converted to use `%w`. Thoughts?

[the remark this answers is not in the document]
Comment (sarken) [reply config/initializers/gem-plugin_config/sanitizer_config.rb:36]: I think it makes sense to match the existing ones, yeah.

File: config/initializers/gem-plugin_config/sanitizer_config.rb [hunk 2]
```diff
@@ -58,5 +58,16 @@ module Config
 
       env[:node]["open"] = "open" if env[:node].has_attribute?("open")
     end
+
+    # On img elements, convert relative paths to absolute:
+    RELATIVE_IMAGE_PATH_TRANSFORMER = lambda do |env|
+      return unless env[:node_name] == "img" && env[:node]["src"]
+
+      base = "https://#{ArchiveConfig.APP_HOST}/"
```
Comment (Bilka2) [inline config/initializers/gem-plugin_config/sanitizer_config.rb:66]: Out of curiosity, why does this use `APP_HOST` instead of `APP_URL`?

**Comment by sarken** [comment]
I can't remember from last time -- does bumping the sanitizer version prompt everything to be resanitized immediately, or does that only happen with a rake task? (james_ notes if we want to resanitize everything, we might need a new queue.)

File: config/initializers/gem-plugin_config/sanitizer_config.rb (hunk 2 above)
Comment (tickinginstant) [reply config/initializers/gem-plugin_config/sanitizer_config.rb]: TBH, I didn't remember it was an option. I have `APP_HOST` overridden in my `local.yml`, so it's on my radar, but `APP_URL` is currently only used in scripts that I don't use, so I've never had to override it.

**Comment by tickinginstant** [comment]
Bumping the sanitizer version will cause anything displayed with `sanitize_field` to be re-sanitized (but not saved to the database anymore). I think the only fields that aren't displayed like that are controlled by admin users, so it's OK to avoid the re-sanitizing task. (In fact, it would be better to avoid the re-sanitizing task for now, because it clobbers the admin log items that use `inspect`.)

**Comment by sarken** [comment]
Hm, yeah, good point about the admin activities. I thought we addressed that, but it's just a display change in #4542. 

Could we have a task to resanitize work and chapter fields that might contain images (I think that's the various notes fields, the summary fields, and the chapter content), though? There's a problem with relative image paths in downloads we'd like to address.

**Comment by sarken** [comment]
(If you want, one of us can add it--I know you're having some issues right now and I don't want to exacerbate them!)

**Comment by tickinginstant** [comment]
The images will be fixed on display, because all of those are displayed with `sanitize_field`:
https://github.com/otwcode/otwarchive/blob/801ce0931c0dbb144f9631eed5e9cf743a6c8d00/app/views/downloads/_download_chapter.html.erb#L12
https://github.com/otwcode/otwarchive/blob/801ce0931c0dbb144f9631eed5e9cf743a6c8d00/app/views/downloads/_download_chapter.html.erb#L18
https://github.com/otwcode/otwarchive/blob/801ce0931c0dbb144f9631eed5e9cf743a6c8d00/app/views/downloads/_download_chapter.html.erb#L30
https://github.com/otwcode/otwarchive/blob/801ce0931c0dbb144f9631eed5e9cf743a6c8d00/app/views/downloads/_download_chapter.html.erb#L37
https://github.com/otwcode/otwarchive/blob/801ce0931c0dbb144f9631eed5e9cf743a6c8d00/app/views/downloads/show.html.erb#L12
https://github.com/otwcode/otwarchive/blob/801ce0931c0dbb144f9631eed5e9cf743a6c8d00/app/views/downloads/_download_afterword.html.erb#L9
https://github.com/otwcode/otwarchive/blob/801ce0931c0dbb144f9631eed5e9cf743a6c8d00/app/views/downloads/_download_preface.html.erb#L52
https://github.com/otwcode/otwarchive/blob/801ce0931c0dbb144f9631eed5e9cf743a6c8d00/app/views/downloads/_download_preface.html.erb#L58

**Review by sarken — APPROVED** [review]
All set then! I'll update the issue so it no longer says we need to resanitize everything. 

Thank you!

**Pull request merged**